File: src/test/java/io/powerrangers/backend/service/CommentServiceTests.java
package io.powerrangers.backend.service;

import io.powerrangers.backend.dto.TaskScope;
import io.powerrangers.backend.dto.TaskStatus;
import io.powerrangers.backend.dto.CommentCreateRequestDto;
import io.powerrangers.backend.dto.CommentResponseDto;
import io.powerrangers.backend.dto.CommentUpdateRequestDto;
import io.powerrangers.backend.entity.Comment;
import io.powerrangers.backend.entity.Task;
import io.powerrangers.backend.entity.User;
import io.powerrangers.backend.dao.CommentRepository;
import io.powerrangers.backend.dao.TaskRepository;
import io.powerrangers.backend.dao.UserRepository;
import io.powerrangers.backend.exception.CustomException;
import jakarta.transaction.Transactional;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import java.time.LocalDateTime;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;

@SpringBootTest
@DisplayName("댓글기능 테스트")
@Transactional
public class CommentServiceTests {

    @Autowired
    CommentService commentService;

    @Autowired
    CommentRepository commentRepository;

    @Autowired
    TaskRepository taskRepository;

    @Autowired
    UserRepository userRepository;

    private User testUser;
    private Task testTask;

    @BeforeEach
    void setUp() {
        testUser = User.builder()
                .nickname("tester")
                .email("tester@naver.com")
                .provider("naver")
                .providerId("naver123")
                .profileImage("http://image.url")
                .build();

        userRepository.save(testUser);

        testTask = Task.builder()
                .category("공부")
                .content("스프링 공부하기")
                .dueDate(LocalDateTime.now().plusDays(3))
                .status(TaskStatus.COMPLETE)
                .taskImage("http://task.img")
                .scope(TaskScope.PUBLIC)
                .user(testUser)
                .build();

        taskRepository.save(testTask);
    }

    @Test
    @DisplayName("최상위 부모 댓글 작성")
    void 최상위댓글() {
        // given
        CommentCreateRequestDto dto = new CommentCreateRequestDto(testTask.getId(), testUser.getId(), null, "최상위 댓글입니다.");

        // when
        commentService.createComment(dto.getUserId(), dto);

        // then
        List<Comment> all = commentRepository.findAll();
        assertThat(all).hasSize(1);
        Comment comment = all.get(0);
        assertThat(comment.getContent()).isEqualTo("최상위 댓글입니다.");
        assertThat(comment.getParent()).isNull();
    }

    @Test
    @DisplayName("기존 부모 댓글이 있는 상태에서 대댓글을 제대로 작성할 수 있는지 테스트")
    void 대댓글() {
        // given
        Comment parent = new Comment(testTask, testUser, null, "부모 댓글입니다.");
        commentRepository.save(parent);

        CommentCreateRequestDto dto = new CommentCreateRequestDto(testTask.getId(), testUser.getId(), parent.getId(), "대댓글입니다");

        // when
        commentService.createComment(dto.getUserId(), dto);

        // then
        List<Comment> comments = commentRepository.findAll();
        assertThat(comments).hasSize(2);

        // 대댓글 찾기
        Comment reply = comments.stream()
                .filter(c -> c.getParent() != null)
                .findFirst()
                .orElseThrow();

        assertThat(reply.getContent()).isEqualTo("대댓글입니다");
        assertThat(reply.getParent().getId()).isEqualTo(parent.getId());
    }

    @Test
    @DisplayName("부모 댓글 있는 상태에서 대댓글2 작성 검증")
    void 대댓글2() {
        //given
        Comment parent = new Comment(testTask, testUser, null, "부모 댓글입니다.");
        commentRepository.save(parent);
        Comment child = new Comment(testTask, testUser, parent, "대댓글입니다");
        commentRepository.save(child);

        CommentCreateRequestDto dto = new CommentCreateRequestDto(testTask.getId(), testUser.getId(), parent.getId(), "대댓글2입니다");

        //when
        commentService.createComment(dto.getUserId(), dto);

        //then
        List<Comment> comments = commentRepository.findAll();
        assertThat(comments).hasSize(3);

        // 대댓글 찾기
        Comment reply = comments.stream()
                .filter(c -> "대댓글2입니다".equals(c.getContent()))
                .findFirst()
                .orElseThrow();

        assertThat(reply.getContent()).isEqualTo("대댓글2입니다");
        assertThat(reply.getParent().getId()).isEqualTo(parent.getId());

        assertThat(child.getParent().getId()).isEqualTo(reply.getParent().getId());
    }

    @Test
    @DisplayName("조회 검증")
    void 조회() {
        Comment parent = new Comment(testTask, testUser, null, "부모 댓글입니다.");
        commentRepository.save(parent);

        Comment child1 = new Comment(testTask, testUser, parent, "대댓글입니다");
        commentRepository.save(child1);

        Comment child2 = new Comment(testTask, testUser, parent, "불라불라");
        commentRepository.save(child2);

        List<CommentResponseDto> result = commentService.getComments(testTask.getId());

        assertThat(result).hasSize(1); // 부모 댓글 하나
        CommentResponseDto parentDto = result.get(0);
        assertThat(parentDto.getContent()).isEqualTo("부모 댓글입니다.");

        assertThat(parentDto.getChildren()).hasSize(2);
        assertThat(parentDto.getChildren())
                .extracting(CommentResponseDto::getContent)
                .containsExactlyInAnyOrder("대댓글입니다", "불라불라");
    }

    @Test
    @DisplayName("댓글 수정 테스트")
    void 댓글수정() {
        Comment parent = new Comment(testTask, testUser, null, "부모댓글");
        commentRepository.save(parent);
        Comment child1 = new Comment(testTask, testUser, parent, "자식댓글");
        commentRepository.save(child1);

        String newContent = "아싸 수정됐다.";
        CommentUpdateRequestDto dto = CommentUpdateRequestDto.builder()
                .content(newContent)
                .build();
        commentService.updateComment(parent.getId(), dto);

        Comment updatedParent = commentRepository.findById(parent.getId())
                .orElseThrow(() -> new RuntimeException("댓글 수정 실패"));

        assertThat(updatedParent.getContent()).isEqualTo(newContent);
    }

    @Test
    @DisplayName("댓글 수정 본인 댓글이 아닌경우")
    void 댓글수정2() {
        /*
         * TODO: 사용자 검증으로 본인의 댓글인지 판단하고 예외를 터트리는 테스트를 만들자 (아직 기능 미구현)
         *
         * */
    }

    @Test
    @DisplayName("댓글 삭제")
    void 댓글삭제() {
        Comment parent = new Comment(testTask, testUser, null, "삭제할 댓글입니다.");
        commentRepository.save(parent);

        commentService.deleteComment(parent.getId());

        boolean exists = commentRepository.existsById(parent.getId());
        assertThat(exists).isFalse();
    }

    @Test
    @DisplayName("댓글 삭제 본인이 아닌 경우")
    void 댓글삭제권한X() {
        Comment parent = new Comment(testTask, testUser, null, "삭제할 댓글입니다.");
        commentRepository.save(parent);

        User attacker = User.builder()
                .nickname("hacker")
                .email("bad@naver.com")
                .provider("naver")
                .providerId("hacker123")
                .profileImage("http://image.url")
                .build();
        userRepository.save(attacker);

        //Todo: 예외부 구현
    }

    @Test
    @DisplayName("댓글 삭제 존재하지 않는 댓글 예외")
    void 삭제시댓글없음() {
        Long nonexistentId = 999L;

        assertThatThrownBy(() -> commentService.deleteComment(nonexistentId))
                .isInstanceOf(CustomException.class);
    }
}

File: src/test/java/io/powerrangers/backend/service/CommentServiceTests.kt
package io.powerrangers.backend.service

import io.mockk.*
import io.powerrangers.backend.dao.CommentRepository
import io.powerrangers.backend.dao.TaskRepository
import io.powerrangers.backend.dao.UserRepository
import io.powerrangers.backend.dto.*
import io.powerrangers.backend.entity.Comment
import io.powerrangers.backend.entity.Task
import io.powerrangers.backend.entity.User
import io.powerrangers.backend.exception.CustomException
import io.powerrangers.backend.exception.ErrorCode
import io.powerrangers.backend.utils.getCurrentUserId
import io.powerrangers.backend.utils.toResponseDto
import io.powerrangers.backend.utils.toUpdateResponseDto
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.assertThrows
import org.springframework.data.repository.findByIdOrNull
import java.time.LocalDateTime
import kotlin.test.Test

class CommentServiceTests {

    val commentRepository = mockk<CommentRepository>()
    val userRepository = mockk<UserRepository>()
    val taskRepository = mockk<TaskRepository>()

    val commentService = CommentService(commentRepository, taskRepository, userRepository)

    val testUser = User(
        id = 1L,
        nickname = "유저1",
        profileImage = null,
        provider = "kakao",
        providerId = "kakao.com",
        email = "kakao@gmail.com",
        role = Role.USER,
    )
    val testTask = Task(
        id = 1L,
        category = "Study",
        content = "Test Comment",
        dueDate = LocalDateTime.now().plusDays(3),
        status = TaskStatus.INCOMPLETE,
        scope = TaskScope.PRIVATE,
        user = testUser
    )

    @BeforeEach
    fun setUp() {
        mockkStatic("io.powerrangers.backend.utils.ContextUtilKt")
        mockkStatic("io.powerrangers.backend.utils.ExtensionsKt")
        every { getCurrentUserId() } returns 1
    }

    @Test
    fun `createComment 성공 시 저장된 댓글 DTO반환`() {
        val request = CommentCreateRequestDto(1L, null, "테스트 댓글")

        val comment = Comment(
            id=1L,
            task = testTask,
            user = testUser,
            parent = null,
            content = "테스트 댓글",
        )
        val expected = CommentResponseDto(
            id = 1L,
            content = "테스트 댓글",
            nickname = "유저1",
            profileImage = null,
            createdAt = LocalDateTime.now(),
            userId = 1L,
            children = emptyList()
        )

        every { commentRepository.save(any()) } returns comment
        every { taskRepository.findByIdOrNull(1L) } returns testTask
        every { userRepository.findByIdOrNull(1L) } returns testUser
        every { commentRepository.findByIdOrNull(1L) } returns null


        every { any<Comment>().toResponseDto(any()) } returns expected

        val result = commentService.createComment(request)

        assertEquals(expected.copy(createdAt = result.createdAt), result)
    }

    @Test
    fun `해당 Task가 존재하지 않을 때 실페테스트`() {
        val request = CommentCreateRequestDto(999L, null, "테스트 댓글")

        every{taskRepository.findByIdOrNull(999L)} returns null
        every{userRepository.findByIdOrNull(1L)} returns testUser

        val exception = assertThrows<CustomException>{
            commentService.createComment(request)
        }

        assertEquals(ErrorCode.TASK_NOT_FOUND.message, exception.errorCode.message)
    }

    @Test
    fun `User가 존재하지 않을 때 실패테스트`() {
        val request = CommentCreateRequestDto(1L, null, "테스트 댓글")

        every{taskRepository.findByIdOrNull(1L)} returns testTask
        every { getCurrentUserId() } returns 999L
        every{userRepository.findByIdOrNull(999L)} returns null

        val exception = assertThrows<CustomException> {
            commentService.createComment(request)
        }

        assertEquals(ErrorCode.USER_NOT_FOUND.message, exception.errorCode.message)
    }

    @Test
    fun `부모댓글이 존재하지 않을 때 실패테스트`() {
        val request = CommentCreateRequestDto(1L,999L,"대댓글")

        every{taskRepository.findByIdOrNull(1L)} returns testTask
        every{userRepository.findByIdOrNull(1L)} returns testUser
        every { commentRepository.findByIdOrNull(999L) } returns null

        val exception = assertThrows<CustomException> {
            commentService.createComment(request)
        }

        assertEquals(ErrorCode.COMMENT_NOT_FOUND.message, exception.errorCode.message)
    }

    @Test
    fun `getComments 성공 시 댓글 반환`() {
        val parentComment = Comment(1L,testTask,testUser,null,"부모 댓글" )
        val childComment = Comment(2L,testTask,testUser,parentComment,"자식댓글")
        parentComment.children.add(childComment)

        every { taskRepository.findByIdOrNull(1L) } returns testTask
        every { commentRepository.findByTaskId(1L) } returns listOf(parentComment, childComment)

        val childDto = CommentResponseDto(
            id = 2L,
            content = "자식댓글",
            nickname = "유저1",
            profileImage = null,
            createdAt = LocalDateTime.now(),
            userId = 1L,
            children = emptyList()
        )

        val parentDto = CommentResponseDto(
            id = 1L,
            content = "부모 댓글",
            nickname = "유저1",
            profileImage = null,
            createdAt = LocalDateTime.now(),
            userId = 1L,
            children = listOf(childDto)
        )

        every { childComment.toResponseDto(any()) } returns childDto
        every { parentComment.toResponseDto(listOf(childDto)) } returns parentDto

        val result = commentService.getComments(1L)

        assertEquals(1, result.size)
        assertEquals("부모 댓글", result[0].content)
        assertEquals(1, result[0].children.size)
        assertEquals("자식댓글", result[0].children[0].content)
    }

    @Test
    fun `getComments 존재하지 않는 Task 실패 테스트`() {
        every { taskRepository.findByIdOrNull(999L) } returns null

        val exception = assertThrows<CustomException> {
            commentService.getComments(999L)
        }

        assertEquals(ErrorCode.TASK_NOT_FOUND.message, exception.errorCode.message)
    }

    @Test
    fun `updateComment 성공 시 수정된 댓글 DTO 반환`() {
        val request = CommentUpdateRequestDto("수정된 댓글")
        val comment = Comment(
            id=1L,
            task = testTask,
            user = testUser,
            parent = null,
            content = "기존 댓글"
        )

        every{commentRepository.findByIdOrNull(1L)} returns comment
        every{comment.toUpdateResponseDto()} returns CommentUpdateResponseDto(id=1L, content = "수정된 댓글", nickname = "유저1", profileImage = null)

        val result = commentService.updateComment(1L,request)

        assertEquals(1L,result.id)
        assertEquals("수정된 댓글",result.content)
    }

    @Test
    fun `updateComment 댓글이 존재하지 않을 시 실패테스트`() {
        val request = CommentUpdateRequestDto("수정된 댓글")

        every { commentRepository.findByIdOrNull(999L) } returns null

        val exception=assertThrows<CustomException> {
            commentService.updateComment(999L,request)
        }

        assertEquals(ErrorCode.COMMENT_NOT_FOUND.message, exception.errorCode.message)
    }

    @Test
    fun `updateComment 댓글의 작성자가 아닐 시 실패테스트`() {
        val request = CommentUpdateRequestDto("수정된 댓글")
        val otherUser = User(
            id=2L,
            nickname="유저2",
            profileImage = null,
            provider = "kakao",
            providerId = "kakao",
            email="kakao2@gmail.com",
            role = Role.USER
            )
        val comment = Comment(
            id = 1L,
            task = testTask,
            user = otherUser,
            parent = null,
            content = "기존 내용"
        )
        every { commentRepository.findByIdOrNull(1L) } returns comment

        val exception = assertThrows<CustomException> {
            commentService.updateComment(1L, request)
        }

        assertEquals(ErrorCode.NOT_THE_OWNER.message, exception.errorCode.message)
    }

    @Test
    fun `deleteComment 성공 시 댓글 삭제`() {
        val comment = Comment(
            id = 1L,
            task = testTask,
            user = testUser,
            parent = null,
            content = "삭제할 댓글"
        )

        every{commentRepository.findByIdOrNull(1L)} returns comment
        every{commentRepository.deleteById(1L)} just Runs

        commentService.deleteComment(1L)

        verify(exactly = 1) {commentRepository.deleteById(1L)}
    }

    @Test
    fun `deleteComment 댓글이 존재하지 않을 시 실페테스트`() {
        every { commentRepository.findByIdOrNull(999L) } returns null

        val exception = assertThrows<CustomException> {
            commentService.deleteComment(999L)
        }

        assertEquals(ErrorCode.COMMENT_NOT_FOUND.message, exception.errorCode.message)
    }

    @Test
    fun `deleteComment 작성자가 아닐 경우 실패테스트`(){
        val otherUser = User(
            id=2L,
            nickname="유저2",
            profileImage = null,
            provider = "kakao",
            providerId = "kakao",
            email="kakao2@gmail.com",
            role = Role.USER
        )
        val comment = Comment(
            id = 1L,
            task = testTask,
            user = otherUser,
            parent = null,
            content = "삭제 시도 댓글"
        )

        every{commentRepository.findByIdOrNull(1L)} returns comment

        val exception = assertThrows<CustomException> {
            commentService.deleteComment(1L)
        }

        assertEquals(ErrorCode.NOT_THE_OWNER.message, exception.errorCode.message)
    }
}
